### PR TITLE
Cleanup

### PR DESCRIPTION
**Pull request by hpihkala**
**Purpose of this PR:**

- Make the util more future-proof by removing most of the stuff that easily gets outdated
- Fix all incorrect/outdated stuff
- Add some useful features like `wait` and `shell` commands (see below)

**Cleanup:**

- Remove interactive mode
- Remove numbered service sets (1-5)
- Update help
- Update and restructure README
- Remove the `--all` modifier (it's now the default)
- Rename `clean` command to `factory-reset` to differentiate it more from `wipe`
- Fix dependencies in docker compose files
- Unify naming in docker compose files
- Add missing health checks to most services

**New features:**

- Add `wait` command to allow scripts to wait for all health checks to pass
- Add `shell` command for easily opening a shell to containers
- Add `wipe` command to wipe persisted data (replaces unexpectedly behaving `-r` flag)
- Add `update` command to update the tool itself
- Add `start --except [service]` option to start all except some services
- Add `start --wait` option to conveniently execute `wait` after `start`

**BREAKING CHANGES:**

Merging this PR will break CI scripts due to removed features/changed syntax. Migration instructions:

- For full stack tests: `stream-docker-dev start`
- For integration tests of a service: `streamr-docker-dev start --except [service-under-test]`

Additionally, if your CI script contains code for waiting for the dockerized services to come up, you can now replace that with simply `streamr-docker-dev wait`.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,23 +1,6 @@
-# Streamr cloud development environment
+# Streamr development environment
 
-This repository contains Docker compose files and command line tool `streamr_docker_dev` for setting up a local Streamr cloud development environment. The cloud environment consists of base services (3rd party) and Streamr services (in-house).
-
-![Streamr cloud architecture](high-level.png)
```
Comment (harbu) [inline README.md]: Might want to remove `high-level.png` and `high-level.xml` files too in this PR.

File: README.md [hunk 2]
```diff
@@ -48,109 +25,120 @@ streamr-docker-dev bind-ip
...
-| Service             | Original localhost url              | New url after `nginx`         |
-|---------------------|-------------------------------------|-------------------------------| 
-| `engine-and-editor` | http://localhost:8081/streamr-core  | http://localhost/streamr-core | 
-| `platform`          | http://localhost:3333               | http://localhost/             | 
+Wipe the persisted data of all services: `streamr-docker-dev wipe`
 
+### `factory-reset`
 
-## Accounts
+"Factory reset" the environment by removing all images and persisted data: `streamr-docker-dev clean`
```
Comment (harbu) [inline README.md]: Command should be `streamr-docker-dev factory-reset`?

File: README.md [hunk 3]
```diff
@@ -48,109 +25,120 @@ streamr-docker-dev bind-ip
...
-streamr-docker-dev <command> --help
+streamr-docker-dev start
+streamr-docker-dev stop [service]
 ```
 
+## Usage in testing and CI
+
+For integration tests of one of the services, do the same as above (stop the docker version of the service under test), then start your tests. In CI, you may want to add some checks to ensure that the services are started before launching your tests.
+
+When running testing the SDKs or running other end-to-end tests, just start the full stack with `streamr-docker-dev start` before running your tests.
```
Comment (harbu) [inline README.md]: <del>running</del>

**Comment by mikhaelsantos** [comment]
If we are going to fix naming would be good to have consistency between '-' or '_'
`streamr_dev_broker-node-no-storage-1`

**Comment by mikhaelsantos** [comment]
Personally I would like to see backward compatibility and a Deprecation warning. 
This is an internal tool used for pretty much everything and everyone has to merge their PR fix at the same time.

File: streamr-docker-dev/bin.sh [hunk 4]
```diff
@@ -3,280 +3,107 @@
...
+            -h | --help )               HELP=1
+                                        ;;
+            -f | --follow )             FOLLOW=1
+                                        ;;
+            --dry-run )                 DRY_RUN=1
+                                        ;;
+            --attached )                DETACHED=0
+                                        ;;
+            * )                         echo "Invalid option: $1"
+                                        exit
```
Comment (kare) [inline streamr-docker-dev/bin.sh]:
```suggestion
                                        exit 1
```

**Comment by hpihkala** [comment]
> If we are going to fix naming would be good to have consistency between '-' or '_'
> `streamr_dev_broker-node-no-storage-1`

@mikhaelsantos Good point. Just to check: Is the purpose of the `streamr_dev_` prefix in container names to avoid clashes with other containers potentially running on the same machine? So it doesn't matter which prefix we choose?

I'll rename all containers to use dashes only, no underscores.

**Comment by hpihkala** [comment]
> ```
> streamr-docker-dev start 5
> Starting 5
> Creating network "streamr-docker-dev_default" with the default driver
> Creating volume "streamr-docker-dev_data" with default driver
> Creating volume "streamr-docker-dev_mysql_init_scripts" with default driver
> Creating volume "streamr-docker-dev_cassandra_init_scripts" with default driver
> ERROR: No such service: 5
> ```
> 
> Shouldn't do anything

@mikhaelsantos I won't be able to address this one; it's beyond the scope of this PR due to the following:
- The `streamr-docker-dev` script only wraps `docker-compose`, and the observed behavior comes directly from there:
```
$ docker-compose up foo
Creating network "streamr-docker-dev_default" with the default driver
Creating volume "streamr-docker-dev_data" with default driver
Creating volume "streamr-docker-dev_mysql_init_scripts" with default driver
Creating volume "streamr-docker-dev_cassandra_init_scripts" with default driver
ERROR: No such service: foo
```
- The same behavior is seen in the `streamr-docker-dev` before my changes
- Looks like it only creates the volumes defined in `docker-compose.yml` on the very first run (after `factory-reset` or on a clean install). Seems like a pretty minor side effect.

**Comment by hpihkala** [comment]
> Personally I would like to see backward compatibility and a Deprecation warning.
> This is an internal tool used for pretty much everything and everyone has to merge their PR fix at the same time.

This is a good point definitely worth discussing. It's a matter of "move fast and break things" vs. "go the extra mile". Even in the latter approach things may still break when the deprecated feature(s) are removed. There's no correct answer, the benefits of both approaches are obvious. Any opinions @harbu @kare?

In practice the breaking changes affect only CI. The changes needed to CI scripts are pretty straight forward.

In my humble personal opinion, in this case I think the best outcome would be achieved if one person updated all the CI scripts, and then merged them in one go once ready. I do understand this does not promote the general goal of each team taking responsibility of their own CI stuff, which I think is an excellent goal in general. But in the time spent discussing, notifying, deprecating, and coordinating, it would already be done so many times over. ;)

**Comment by kare** [comment]
@hpihkala I'd go for "move fast and break things". I don't think that the Streamr developer organisation is large enough to require a deprecation warning at the moment. I think it would be great to have streamr-docker-dev versioned. It guess it wouldn't help immediately with this case, but in the future in case of backwards incompatible update we could update project at a time or even in (rare) case of a bug in streamr-docker-dev we could refer to an earlier working version.

cc: @harbu @mikhaelsantos

**Pull request merged**